File: pigain/nodes/gp.py
import numpy as np

from collections import namedtuple
HyperParam = namedtuple("HyperParam", "l sigma_f sigma_n")

def sqexpkernel(x1, x2, hyperparam):
    n1 = x1.shape[0]
    n2 = x2.shape[0]
    K = np.empty([n1, n2])
    # print(n1, n2)

    for i in range(0,n2):
        # print("Hej")
        # print(x1)
        # print(x2[i,:])
        # print(x1 - x2[i,:])
        l = np.linalg.norm(x1 - x2[i,:], axis = 1)
        # print(l)
        # print(hyperparam.sigma_f**2 * np.exp(-0.5 * (l / hyperparam.l)**2))
        # print("BYE")
        K[:,i] = hyperparam.sigma_f**2 * np.exp(-0.5 * (l / hyperparam.l)**2)

    return K

def gp(y, x, xstar, hyperparam, kernel):
    if(y.shape[0] is 0 or x.shape[0] is 0):
        return (np.empty((0)), np.empty((0)))

    K     = kernel(x, x, hyperparam)            # K(x,x)
    kstar = kernel(x, xstar, hyperparam)        # K(x,x*)
    kss   = kernel(xstar, xstar, hyperparam)    # K(x*,x*)

    # Algorithm 2.1 from Rasmussen & Williams
    L = np.linalg.cholesky(K + hyperparam.sigma_n**2*np.identity(K.shape[0]))
    alpha = np.linalg.solve( np.transpose(L), np.linalg.solve(L, y))
    posterior_mean = np.matmul(np.transpose(kstar), alpha)

    v = np.linalg.solve(L, kstar)
    posterior_variance = np.diag(kss - np.matmul(np.transpose(v), v))

    return posterior_mean, posterior_variance

File: pigain/src/gp.cpp
#include <pigain/gp.h>

#include <eigen3/Eigen/Cholesky>

#include <omp.h>

#include <ros/ros.h>

namespace pig
{
Eigen::MatrixXd sqExpKernel(const Eigen::Matrix<double, Eigen::Dynamic, 3>& x1,
                            const Eigen::Matrix<double, Eigen::Dynamic, 3>& x2, double hyp_l, double hyp_sigma_f,
                            double hyp_sigma_n)
{
  size_t n1 = x1.rows();
  size_t n2 = x2.rows();
  Eigen::MatrixXd k(n1, n2);

  // #pragma omp parallel for
  for (size_t i = 0; i < n2; ++i)
  {
    // Eigen::Matrix<double, Eigen::Dynamic, 3> temp = x1 - x2.row(i);

    // FIXME: Optimize (find cool Eigen functions)
    for (size_t j = 0; j < n1; ++j)
    {
      // double l = temp.row(j).norm();
      double l = (x1.row(j) - x2.row(i)).norm();

      k(j, i) = std::pow(hyp_sigma_f, 2) * std::exp(-0.5 * std::pow(l / hyp_l, 2));
    }
  }

  return k;
}

std::pair<double, double> gp(const Eigen::VectorXd& y, const Eigen::Matrix<double, Eigen::Dynamic, 3>& x,
                             const Eigen::Matrix<double, Eigen::Dynamic, 3>& x_star, double hyp_l, double hyp_sigma_f,
                             double hyp_sigma_n)
{
  if (y.size() == 0 || x.size() == 0)
  {
    return std::make_pair(0, 0);
  }

  Eigen::MatrixXd k = sqExpKernel(x, x, hyp_l, hyp_sigma_f, hyp_sigma_n);
  Eigen::MatrixXd k_star = sqExpKernel(x, x_star, hyp_l, hyp_sigma_f, hyp_sigma_n);
  Eigen::MatrixXd k_star_star = sqExpKernel(x_star, x_star, hyp_l, hyp_sigma_f, hyp_sigma_n);

  // Algorithm 2.1 from Rasmussen & Williams
  // FIXME: Check that everything is correct and optmize (find cool Eigen functions)
  Eigen::LLT<Eigen::MatrixXd> l(k + std::pow(hyp_sigma_n, 2) * Eigen::MatrixXd::Identity(k.rows(), k.cols()));
  Eigen::MatrixXd alpha = l.matrixL().transpose().solve(l.matrixL().solve(y));

  double posterior_mean = (k_star.transpose() * alpha)(0, 0);

  Eigen::MatrixXd v = l.matrixL().solve(k_star);
  double posterior_variance = (k_star_star - (v.transpose() * v)).diagonal()(0, 0);

  return std::make_pair(posterior_mean, posterior_variance);
}
}  // namespace pig